quota/quota.go

```diff
@@ -0,0 +1,180 @@
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
+
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
+
+	// parentGroup is the the parent group that this group is a child of. It is
+	// here mainly for coding convenience to easily construct the full chain
+	// back to a root group to generate the full slice unit name of the quota
+	// group. If it is nil, then this is a "root" quota group.
+	parentGroup *Group
+}
+
+// MemoryExhaustionBehavior is the behavior determining what the system should
+// do when memory available to a quota group is exhausted.
+type MemoryExhaustionBehavior int
+
+const (
+	// DefaultOOMKiller is the default behavior to invoke to choose which
+	// processes when the memory quota resource limit is reached, and it
+	// consists of letting the Linux kernel's oom-killer run and decide, which
+	// is available on all cgroups version systems, but is not deterministic as
+	// to which processes are killed.
+	DefaultOOMKiller MemoryExhaustionBehavior = iota
+)
+
+// SliceFileName returns the name of the slice file that should be used for this
+// quota group. This name will include all of the group's parents in the name.
+// For example, a group named "bar" that is a child of the "foo" group will have
+// a systemd slice name as "foo-bar.slice". Note that the slice name may differ
+// from the snapd friendly group name, mainly in the case that the group is a
+// sub group.
+func (grp *Group) SliceFileName() string {
+	if grp.parentGroup == nil {
+		// root group name, then the slice unit is just "<name>.slice"
+		return fmt.Sprintf("%s.slice", grp.Name)
+	}
+
+	// otherwise we need to track back to get all of the parent elements
+	grpNames := []string{}
+	parentGrp := grp.parentGroup
+	for parentGrp != nil {
+		grpNames = append([]string{parentGrp.Name}, grpNames...)
+		parentGrp = parentGrp.parentGroup
+	}
+
+	buf := &bytes.Buffer{}
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
+	}
+	fmt.Fprintf(buf, "%s.slice", grp.Name)
+	return buf.String()
+}
+
+// allow only alphanumerics at least one character long, and not starting with
+// a number
+var validGroupNameRegexp = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9]*$`)
+
+func (grp *Group) validate() error {
+	// check that the name is a simple alphanumeric name
+	if !validGroupNameRegexp.MatchString(grp.Name) {
+		return fmt.Errorf("group name %q contains invalid characters (valid names are alphanumeric starting with a letter)", grp.Name)
+	}
+
+	// check if the name is reserved for future usage
+	switch grp.Name {
+	case "root", "system", "snapd", "user":
```

Comment:
so could one play some name squatting and use a group user1000, which produces user1000.slice, very similar to user-1000.slice which is crated for a real user?

Reply:
Yes that issue does currently exist, we could enforce that names need to not even start with these prefixes but I don't think it's necessary given that we will prefix the real slice name with `snap.` and that systemd doesn't use anything like `user1000.slice` currently, it's only with `user-1000.slice` or `user@1000.slice` etc